airbyte-integrations/connectors/source-microsoft-sharepoint/source_microsoft_sharepoint/stream_reader.py

```diff
@@ -0,0 +1,197 @@
+        else:
+            result = self.msal_app.acquire_token_for_client(scopes=scope)
+
+        if "access_token" not in result:
+            error_description = result.get("error_description", "No error description provided.")
+            raise MsalServiceError(error=result.get("error"), error_description=error_description)
+
+        return result
+
+
+class SourceMicrosoftSharePointStreamReader(AbstractFileBasedStreamReader):
+    """
+    A stream reader for Microsoft SharePoint. Handles file enumeration and reading from SharePoint.
+    """
+
+    ROOT_PATH = [".", "/"]
+
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def config(self) -> SourceMicrosoftSharePointSpec:
+        return self._config
+
+    @property
+    def one_drive_client(self) -> SourceMicrosoftSharePointSpec:
+        return SourceMicrosoftSharePointClient(self._config).client
+
+    @config.setter
+    def config(self, value: SourceMicrosoftSharePointSpec):
+        """
+        The FileBasedSource reads and parses configuration from a file, then sets this configuration in its StreamReader. While it only
+        uses keys from its abstract configuration, concrete StreamReader implementations may need additional keys for third-party
+        authentication. Therefore, subclasses of AbstractFileBasedStreamReader should verify that the value in their config setter
+        matches the expected config type for their StreamReader.
+        """
+        assert isinstance(value, SourceMicrosoftSharePointSpec)
+        self._config = value
+
+    def list_directories_and_files(self, root_folder, path=None):
+        """Enumerates folders and files starting from a root folder."""
+        drive_items = execute_query_with_retry(root_folder.children.get())
+        found_items = []
+        for item in drive_items:
+            item_path = path + "/" + item.name if path else item.name
+            if item.is_file:
+                found_items.append((item, item_path))
+            else:
+                found_items.extend(self.list_directories_and_files(item, item_path))
+        return found_items
+
+    def get_files_by_drive_name(self, drives, folder_path):
+        """Yields files from the specified drive."""
+        path_levels = [level for level in folder_path.split("/") if level]
+        folder_path = "/".join(path_levels)
+
+        for drive in drives:
+            is_sharepoint = drive.drive_type == "documentLibrary"
+            if is_sharepoint:
+                folder = (
+                    drive.root if folder_path in self.ROOT_PATH else execute_query_with_retry(drive.root.get_by_path(folder_path).get())
+                )
+                yield from self.list_directories_and_files(folder)
+
+    @property
+    @lru_cache(maxsize=None)
+    def drives(self):
+        """
+        Retrieves and caches SharePoint drives, including the user's drive based on authentication type.
```

Comment:
Could you please provide an example? I'm following the guidelines outlined in [the API docs](https://learn.microsoft.com/en-us/sharepoint/list-onedrive-urls) and [the library docs](https://github.com/vgrem/Office365-REST-Python-Client/blob/master/examples/onedrive/drives/list.py), which suggest this approach.

Reply:
@maxi297, I've created [an issue](https://github.com/airbytehq/airbyte/issues/34768) for further investigation since I'm currently unable to determine how to distinguish between SharePoint and OneDrive shared drives.